### ISSUE TITLE
Latent GP with random effects?

### ISSUE THREAD

**Issue by fonnesbeck** [cornellius-gp/gpytorch#1226]
Is it currently possible to combine GPs from GPyTorch with more general Bayesian modeling? Specifically, I'm thinking of latent Gaussian processes that are combined with things like a hierarchical random effect; you might want a GP to model changes in some parameter over time, but have individuals have their own means, for example. Would this be possible by integrating with Pyro as has been done in some of the examples? Pointers to examples, if any, would be helpful.

**Comment by gpleiss** [comment]
I believe what you are trying to model **should** be possible with GPyTorch + Pyro, but we have no examples currently for this type of model. Could you provide some more information? (Maybe a link to a paper or some sort of illustration of the model?)

**Comment by fonnesbeck** [comment]
Ok, thanks. I will see if I can cobble something together.

What I'm trying to model is a time-varying baseline parameter, for which I am using the GP:

f ~ GP(m, K)

Individuals for which observations are measured have a range of ability, so an individual random effect is appropriate:

m_i ~ N(mu, tau)

resulting in an expected value for an individual at any given time:

theta_it = m_i + f[t]

which ends up in the likelihood of observed values:

y_jit ~ N(theta_it, sigma)

I can build this in PyMC3, but the GPs are slow for the amount of data I have, and was looking to use the fast BBMM implementation that you have.

**Comment by gpleiss** [comment]
Okay gotcha. Given that everything is Gaussian (yay!) exact inference should be possible. It might be best not to combine this with Pyro then.

What are `mu` and `tau` in this scenario? Are they learnable parameters? One option would be to create a GP model class that directly produces `theta_it`.

```python
class CombinedIndividualAndTimeGP(gpytorch.models.ExactGP):
    def __init__(self, train_time, train_person_indices, train_targets, num_people):
        likelihood = gpytorch.likelihoods.GaussianLikelihood()
        
        super().__init__(
            train_inputs=(train_targets, train_person_indices),
            train_targets=train_targets,
            likelihood=likelihood
        )
        
        # Parameters for individual effects
        self.mu = torch.nn.Parameter(torch.randn(num_people))
        self.tau = torch.nn.Parameter(torch.randn(num_people))
        
        # Parameters for timing
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.MaternKernel(nu=2.5)
        
    def forward(self, times, person_indices):
        time_mean = self.mean_module(times)
        time_covar = self.covar_module(times).evaluate_kernel()
        person_mean = self.mu[person_indices.squeeze(-1)]
        person_covar = gpytorch.lazy.DiagLazyTensor(self.tau[person_indices.squeeze(-1)].pow(2.))
        
        return gpytorch.distributions.MultivariateNormal(
            time_mean + person_mean,
            time_covar + person_covar
        )
        

# How to use it
model = CombinedIndividualAndTimeGP(torch.linspace(0., 1., 20), torch.zeros(20, dtype=torch.long), torch.randn(20), 1)
model.eval()
model(torch.linspace(0., 1., 20), torch.zeros(20, dtype=torch.long))
```

**Comment by fonnesbeck** [comment]
Yeah, mu and tau are hierarchical parameters that describe the expected value and sd of the distribution of individual random effects. The idea is there is individual variation as well as the temporal, and the values of m_i are these individual effects.

**Comment by gpleiss** [comment]
Cool! Hopefully this model is a good starting point for you.

**Comment by fonnesbeck** [comment]
Will use it as a starting point, thanks!

**Comment by fonnesbeck** [comment]
The key piece that is missing is that there does not appear to be a distribution of individuals here. It looks like there are N individual mu's and tau's, when in fact there should be a single mu and a single tau that, in turn, describe the distribution of individual values. So, there needs to be another Gaussian in there that describes the "distribution of individuals" (which if different than the distribution of data in the likelihood). Is this where pyro might be needed?

**Comment by gpleiss** [comment]
Ah I see. Unless I am misunderstanding something, I think the code example that you provide should work.

**Comment by sommukh** [comment]
I have a small query related to this approach of incorporating individual specific random effects along with a GP. How do we pass the extra parameters (i.e. train_person_indices, train_targets, num_people) to the evaluation code while making predictions?

` predictions = likelihood(model(X_test)) <-- I guess I need to pass the extra parameters here, but I am not sure how`
`mean = predictions.mean`
`lower, upper = predictions.confidence_region()`

Kindly help me

**Comment by gpleiss** [comment]
> How do we pass the extra parameters (i.e. train_person_indices, train_targets, num_people) to the evaluation code while making predictions?

I'm not sure what you mean here. Where does `train_person_indices` and `num_people` occur in your code example?

**Comment by sommukh** [comment]
> > How do we pass the extra parameters (i.e. train_person_indices, train_targets, num_people) to the evaluation code while making predictions?
> 
> I'm not sure what you mean here. Where does `train_person_indices` and `num_people` occur in your code example?

@gpleiss  ... thank you for your response, I think my question is not very clear ... I shall put it as a separate question with more illustrations ...
